src/Microsoft.AspNetCore.Blazor.Browser/Http/BrowserHttpMessageHandler.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Blazor.Browser.Interop;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Blazor.Browser.Http
+{
+    /// <summary>
+    /// A browser-compatible implementation of <see cref="HttpMessageHandler"/>
+    /// </summary>
+    public class BrowserHttpMessageHandler : HttpMessageHandler
+    {
+        static object _idLock = new object();
+        static int _nextRequestId = 0;
+        static IDictionary<int, TaskCompletionSource<HttpResponseMessage>> _pendingRequests
+            = new Dictionary<int, TaskCompletionSource<HttpResponseMessage>>();
+
+        /// <inheritdoc />
+        protected override async Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSource<HttpResponseMessage>();
+            cancellationToken.Register(() => tcs.TrySetCanceled());
+
+            int id;
+            lock (_idLock)
+            {
+                id = _nextRequestId++;
+                _pendingRequests.Add(id, tcs);
+            }
+
+            RegisteredFunction.Invoke<object>(
+                $"{typeof(BrowserHttpMessageHandler).FullName}.Send",
+                id,
+                request.Method.Method,
+                request.RequestUri,
+                request.Content == null ? null : await GetContentAsString(request.Content),
+                SerializeHeadersAsJson(request));
+
+            return await tcs.Task;
+        }
+
+        private string SerializeHeadersAsJson(HttpRequestMessage request)
+            => JsonUtil.Serialize(
+                (from header in request.Headers.Concat(request.Content?.Headers ?? Enumerable.Empty<KeyValuePair<string, IEnumerable<string>>>())
+                 from headerValue in header.Value // There can be more than one value for each name
+                 select new[] { header.Key, headerValue }).ToList()
+            );
+
+        private static async Task<string> GetContentAsString(HttpContent content)
+            => content is StringContent stringContent
+                ? await stringContent.ReadAsStringAsync()
+                : throw new InvalidOperationException($"Currently, {typeof(HttpClient).FullName} " +
```

Comment:
I hope this will be updated to support any `HttpContent` later as this is very limiting.

Reply:
If you'd be interested in submitting a PR for it, we could certainly have it done sooner :)

Reply:
I'm looking into this and am wondering what's the reason you can't invoke `ReadAsStringAsync` on any `HttpContent`?